File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilLefarmaCrawler.java
package br.com.lett.crawlernode.crawlers.corecontent.brasil;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.json.JSONObject;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import br.com.lett.crawlernode.core.fetcher.models.Request;
import br.com.lett.crawlernode.core.fetcher.models.Request.RequestBuilder;
import br.com.lett.crawlernode.core.models.Card;
import br.com.lett.crawlernode.core.models.CategoryCollection;
import br.com.lett.crawlernode.core.models.Product;
import br.com.lett.crawlernode.core.models.ProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.Crawler;
import br.com.lett.crawlernode.util.CrawlerUtils;
import br.com.lett.crawlernode.util.Logging;
import models.prices.Prices;

public class BrasilLefarmaCrawler extends Crawler {

  public BrasilLefarmaCrawler(Session session) {
    super(session);
  }

  @Override
  public List<Product> extractInformation(Document doc) throws Exception {
    super.extractInformation(doc);
    JSONObject fetchedJson = new JSONObject();
    List<Product> products = new ArrayList<>();

    if (isProductPage(doc)) {
      Logging.printLogDebug(logger, session, "Product page identified: " + this.session.getOriginalURL());

      JSONObject script = CrawlerUtils.selectJsonFromHtml(doc, "script", "dataLayer[0]['product'] = ", ";", false, true);
      if (script.has("id")) {
        String url = "https://www.lefarma.com.br/public_api/v1/products/" + script.get("id");

        Request request = RequestBuilder.create().setUrl(url).setCookies(cookies).build();
        fetchedJson = CrawlerUtils.stringToJson(this.dataFetcher.get(session, request).getBody());
      }

      String internalPid = crawlInternalPid(fetchedJson);
      String mainName = fetchedJson.has("name") ? fetchedJson.get("name").toString().trim() : null;
      Float price = CrawlerUtils.scrapSimplePriceFloat(doc, "span[itemprop=price]", false);
      String description = crawlDescription(fetchedJson);
      CategoryCollection categories = crawlCategories(fetchedJson);

      for (Object obj : fetchedJson.getJSONArray("variants")) {

        JSONObject jsonObject = (JSONObject) obj;

        String name = crawlName(jsonObject);

        if (mainName != null && !mainName.toLowerCase().contains(name.toLowerCase())) {
          name = mainName + " " + name;
        }

        String internalId = crawlInternalId(jsonObject);
        String primaryImage = crawlPrimaryImage(jsonObject);
        boolean available = crawlAvailability(jsonObject);
        Prices prices = crawlPrices(jsonObject);

        Product product = ProductBuilder.create().setUrl(session.getOriginalURL()).setInternalId(internalId).setInternalPid(internalPid).setName(name)
            .setPrice(price).setPrices(prices).setAvailable(available).setCategory1(categories.getCategory(0)).setCategory2(categories.getCategory(1))
            .setCategory3(categories.getCategory(2)).setPrimaryImage(primaryImage).setSecondaryImages(null).setDescription(description)
            .setMarketplace(null).build();

        products.add(product);

      }
    } else {
      Logging.printLogDebug(logger, session, "Not a product page " + this.session.getOriginalURL());
    }

    return products;

  }

  private String crawlDescription(JSONObject obj) {
    String descprition = null;

    if (obj.has("htmlDescriptions")) {
      for (Object objDescriptions : obj.getJSONArray("htmlDescriptions")) {
        JSONObject jsonObject = (JSONObject) objDescriptions;
        if (jsonObject.has("description")) {
          descprition = jsonObject.getString("description");
        }
      }
    }

    return descprition;
  }

  private CategoryCollection crawlCategories(JSONObject obj) {
    CategoryCollection categories = new CategoryCollection();
    if (obj.has("breadcrumbs")) {
      for (Object category : obj.getJSONArray("breadcrumbs")) {
        JSONObject jsonObjCategory = (JSONObject) category;
        if (jsonObjCategory.has("name")) {
          categories.add(jsonObjCategory.getString("name"));
        }
      }
    }
    return categories;
  }

  private String crawlName(JSONObject obj) {
    String name = null;

    if (obj.has("definition1Value")) {
      name = obj.getString("definition1Value").trim();
    }

    return name;
  }

  private String crawlInternalId(JSONObject obj) {
    String internalId = null;

    if (obj.has("id")) {
      internalId = obj.get("id").toString();
    }

    return internalId;
  }

  private String crawlInternalPid(JSONObject obj) {
    String internalPid = null;

    if (obj.has("erpId")) {
      internalPid = obj.getString("erpId");
    }

    return internalPid;
  }


  private String crawlPrimaryImage(JSONObject obj) {
    String primaryImage = null;

    if (obj.has("mainImage")) {
      primaryImage = obj.getString("mainImage");
    }

    return primaryImage;
  }

  private boolean crawlAvailability(JSONObject obj) {
    return obj.has("available") ? obj.getBoolean("available") : false;
  }

  private Prices crawlPrices(JSONObject obj) {
    Prices prices = new Prices();
    Map<Integer, Float> installmentPriceMap = new TreeMap<>();

    if (!obj.isNull("discountValue")) {
      if (obj.has("promotionPrice")) {
        prices.setBankTicketPrice(CrawlerUtils.getDoubleValueFromJSON(obj, "promotionPrice"));
        prices.setPriceFrom(CrawlerUtils.getDoubleValueFromJSON(obj, "price"));
      }
    } else {
      prices.setBankTicketPrice(CrawlerUtils.getDoubleValueFromJSON(obj, "price"));
    }

    if (obj.has("hasInstallmentsWithInterest") && obj.getBoolean("hasInstallmentsWithInterest")) {

      if (obj.has("quantityOfInstallmentsWithInterest") && !obj.isNull("quantityOfInstallmentsWithInterest")) {
        installmentPriceMap.put(obj.getInt("quantityOfInstallmentsWithInterest"),
            CrawlerUtils.getFloatValueFromJSON(obj, "valueOfInstallmentsWithInterest"));
      }

    } else {

      if (obj.has("quantityOfInstallmentsNoInterest") && !obj.isNull("quantityOfInstallmentsNoInterest")) {
        installmentPriceMap.put(obj.getInt("quantityOfInstallmentsNoInterest"),
            CrawlerUtils.getFloatValueFromJSON(obj, "valueOfInstallmentsNoInterest"));
      }

    }
    prices.insertCardInstallment(Card.SHOP_CARD.toString(), installmentPriceMap);

    return prices;
  }



  private boolean isProductPage(Element e) {
    return e.selectFirst(".produto_comprar") != null;
  }
}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilLefarmaCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.brasil

import br.com.lett.crawlernode.core.models.Card
import br.com.lett.crawlernode.core.models.Product
import br.com.lett.crawlernode.core.models.ProductBuilder
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.Crawler
import br.com.lett.crawlernode.util.*
import models.Offer.OfferBuilder
import models.Offers
import models.pricing.Installment.InstallmentBuilder
import models.pricing.Installments
import models.pricing.Pricing.PricingBuilder
import org.jsoup.nodes.Document
import java.util.*

class BrasilLefarmaCrawler(session: Session?) : Crawler(session) {

   override fun extractInformation(document: Document): List<Product> {
      val products = mutableListOf<Product>()
      if (document.selectFirst(".nome-produto.titulo.cor-secundaria") != null) {

         val description = CrawlerUtils.scrapSimpleDescription(document, Arrays.asList("#descricao"))
         val name = CrawlerUtils.scrapStringSimpleInfo(document,".nome-produto.titulo",true)
         val internalId = CrawlerUtils.scrapStringSimpleInfoByAttribute(document,"div[data-trustvox-product-code-js]","data-trustvox-product-code-js")
         val primaryImage = CrawlerUtils.scrapSimplePrimaryImage(document,"#imagemProduto",Arrays.asList("src"),"https:","cdn.awsli.com.br")
         val available = document.selectFirst(".comprar a.botao.botao-comprar") != null
         val offers = if (available) scrapOffers(document) else Offers()
         val categories = document.select(".info-principal-produto ul a")?.eachText(ignoreIndexes = arrayOf(0))
         val secondaryImages = document.select(".miniaturas.slides a")?.toSecondaryImagesBy(attr = "data-imagem-grande", ignoreIndexes = arrayOf(0))
         val product = ProductBuilder.create()
            .setDescription(description)
            .setCategories(categories)
            .setPrimaryImage(primaryImage)
            .setSecondaryImages(secondaryImages)
            .setOffers(offers)
            .setUrl(session.originalURL)
         if (document.selectFirst(".atributo-item") != null) {
            for (elem in document.select(".atributo-item")) {
               products.add(
                  product.setInternalId(elem.attr("data-variacao-id"))
                     .setName("$name ${elem.selectFirst(".atributo-item span")?.text()?.trim() ?: ""}".trim()).build()
               )
            }
         } else {
            product.setInternalId(internalId)
               .setName(name)
               .build()
            products.add(product.build())
         }
      }

      return products
   }

   private fun scrapOffers(document: Document): Offers {
      val offers = Offers()
      val price = document.selectFirst(".preco-promocional.cor-principal")?.toDoubleComma()
      val priceFrom = document.selectFirst(".principal .preco-venda")?.toDoubleComma()

      val installments = Installments()
      if (document.selectFirst(".parcela") != null) {
         for (elem in document.select(".parcela")) {
            val instNumber = elem.selectFirst(".cor-principal")?.toInt()
            val instVal = elem.text()?.substringAfter("R$")?.toDoubleComma()
            installments.add(
               InstallmentBuilder.create()
                  .setInstallmentNumber(instNumber)
                  .setInstallmentPrice(instVal).build()
            )
         }
      } else {
         installments.add(
            InstallmentBuilder.create()
               .setInstallmentNumber(1)
               .setInstallmentPrice(price).build()
         )
      }

      val bankSlip = document.selectFirst(".text-parcelas.pull-right.cor-principal")?.toDoubleComma()?.toBankSlip()

      val creditCards = setOf(Card.VISA, Card.MASTERCARD, Card.ELO, Card.AMEX, Card.HIPERCARD, Card.AURA).toCreditCards(installments)

      val pricing = PricingBuilder.create()
         .setSpotlightPrice(price)
         .setPriceFrom(priceFrom)
         .setBankSlip(bankSlip)
         .setCreditCards(creditCards)
         .build()

      offers.add(
         OfferBuilder.create()
            .setPricing(pricing)
            .setSellerFullName("Le Farma")
            .setUseSlugNameAsInternalSellerId(true)
            .setIsMainRetailer(true)
            .setIsBuybox(false)
            .build()
      )

      return offers
   }
}
